header/headertest/fraud/testing.go

```diff
@@ -0,0 +1,100 @@
+package headerfraud
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/ipfs/go-blockservice"
+	"github.com/stretchr/testify/require"
+	"github.com/tendermint/tendermint/libs/bytes"
+	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
+	"github.com/tendermint/tendermint/types"
+
+	"github.com/celestiaorg/celestia-app/pkg/da"
+	"github.com/celestiaorg/nmt"
+	"github.com/celestiaorg/rsmt2d"
+
+	"github.com/celestiaorg/celestia-node/header"
+	"github.com/celestiaorg/celestia-node/header/headertest"
+	"github.com/celestiaorg/celestia-node/share/eds"
+	"github.com/celestiaorg/celestia-node/share/eds/edstest"
+	"github.com/celestiaorg/celestia-node/share/ipld"
+)
+
+// FraudMaker allows to produce an invalid header at the specified height in order to produce the
+// BEFP.
+type FraudMaker struct {
+	t *testing.T
+
+	vals   []types.PrivValidator
+	valSet *types.ValidatorSet
+
+	// height of the invalid header
+	height int64
+
+	prevHash bytes.HexBytes
+}
+
+func NewFraudMaker(t *testing.T, height int64, vals []types.PrivValidator, valSet *types.ValidatorSet) *FraudMaker {
+	return &FraudMaker{
+		t:      t,
+		vals:   vals,
+		valSet: valSet,
+		height: height,
+	}
+}
+
+func (f *FraudMaker) MakeExtendedHeader(odsSize int, edsStore *eds.Store) header.ConstructFn {
+	return func(h *types.Header,
+		comm *types.Commit,
+		vals *types.ValidatorSet,
+		eds *rsmt2d.ExtendedDataSquare,
+	) (*header.ExtendedHeader, error) {
+		if h.Height < f.height {
+			return header.MakeExtendedHeader(h, comm, vals, eds)
+		}
+
+		hdr := *h
+		if h.Height == f.height {
+			adder := ipld.NewProofsAdder(odsSize)
+			square := edstest.RandByzantineEDS(f.t, odsSize, nmt.NodeVisitor(adder.VisitFn()))
```

Comment:
I'm not sure if this is the right place to test such a thing, something closer to a unit test might be a better spot, but do we (or are we) testing for BEFP generation for squares that have out of order blobs?

Reply:
To test out-of-order blobs we need https://github.com/celestiaorg/celestia-node/pull/2585 to be merged, because now [extendBlock](https://github.com/celestiaorg/celestia-node/blob/9d663cf7d948fefa961d9b3c8d0afb8dca3ec8ab/core/eds.go#L26) uses ComputeExtendedDataSquare that will fail in this case.

Reply:
> I'm not sure if this is the right place to test such a thing

This is not the place where we test them. This is a helper function that creates an invalid header. So I decided to put it near functions that creates other test headers